### PR TITLE
Proposed changes to the mkmmd loss function

### PR DESCRIPTION
**Pull request by emersodb**
This is a proposed refactor to the mkmmd loss function constructions. There are a lot of tests here and I can definitely improve the documentation. However, I wanted to get it in front of you ASAP.

Note: Most of the other files are miscellaneous stuff that was failing for me in mypy, black, or flake.

### CODE REVIEW

File: fl4health/losses/mkmmd_loss.py [hunk 1]
```diff
@@ -2,153 +2,202 @@
...
+        # all_h_u_per_vi has shape number kernels x number of v_is. Each row is h_u(v_i) for kernel u with the ith
+        # column being that kernel evaluated on v_i.
+        # output shape is number of kernels x 1
+        return torch.mean(all_h_u_per_v_i, dim=1, keepdim=True)
+
+    def compute_mkmmd(self, X: torch.Tensor, Y: torch.Tensor, beta: torch.Tensor) -> torch.Tensor:
+        # In this function, we assume that X, Y: n_samples, n_features are the same size and that beta is a tensor of
+        # shape number of kernels x 1
+        all_h_u_per_vi = self.compute_all_h_u_per_v_i(X, Y)
+        hat_d_per_kernel = self.compute_hat_d_per_kernel(all_h_u_per_vi)
```
Comment (sanaAyrml) [inline fl4health/losses/mkmmd_loss.py:128]: Do we want to also compute estimated hat_d for mlmmd computation? In my original code I basically computed over all pairs and only used estimation (pairing i and i+1) for optimizing beta.
```
def compute_mmd(
        self,
        XX: torch.Tensor,
        YY: torch.Tensor,
        YX: torch.Tensor,
        XY: torch.Tensor,
    ) -> torch.Tensor:

        return XX.mean(dim=(1, 2)) + YY.mean(dim=(1, 2)) - XY.mean(dim=(1, 2)) - YX.mean(dim=(1, 2))
```

File: fl4health/losses/mkmmd_loss.py [hunk 2]
```diff
@@ -167,12 +216,50 @@ def forward(self, Xs: torch.Tensor, Xt: torch.Tensor) -> torch.Tensor:
...
+        # check to see that at least one of them is positive. If none of them are positive, then select a single kernel
+        # with largest hat_d, similar to the suggestion of Gretton et al. in "Optimal Kernel Choice for Large-Scale
+        # Two-Sample Tests", 2012
+        if not torch.any(hat_d_per_kernel > 0):
+            log(
+                INFO,
+                f"None of the estimates for hat_d are positive: {hat_d_per_kernel}. Rather than optimizing, we select "
+                "a single kernel with largest hat_d_k",
+            )
+            return self.beta_with_largest_hat_d(hat_d_per_kernel)
```
Comment (sanaAyrml) [inline fl4health/losses/mkmmd_loss.py:239]: shouldn't we return beta with smallest hat d if we want to maximize the distance between two distributions? (minimize_type_two_error=False)

File: fl4health/losses/mkmmd_loss.py [hunk 3]
```diff
@@ -167,12 +216,50 @@ def forward(self, Xs: torch.Tensor, Xt: torch.Tensor) -> torch.Tensor:
...
         #   h is (kernel_num, 1)
         #   A is (1, kernel_num)
-        #   b is ()
-        beta = self.sign * QPFunction(verbose=False)(Q, p, G, h, A, b)
+        #   b is a scalar
+
+        # QPFunction returns betas in the shape 1 x num_kernels to we take the transpose for consistency
+        return QPFunction(verbose=False, solver=QPSolvers.CVXPY)(regularized_Q_k, p, G, h, hat_d_per_kernel.t(), b).t()
+
+    def optimize_betas(self, X: torch.Tensor, Y: torch.Tensor, lambda_m: float) -> torch.Tensor:
```
Comment (sanaAyrml) [inline fl4health/losses/mkmmd_loss.py:224]: Maybe have a default value of 1e-5 for  lambda_m?

File: fl4health/losses/mkmmd_loss.py (hunk 2 above)
Comment (sanaAyrml) [reply fl4health/losses/mkmmd_loss.py:239]: or maybe we should first the condition `if self.minimize_type_two_error` ?

File: fl4health/losses/mkmmd_loss.py (hunk 1 above)
Comment (emersodb) [reply fl4health/losses/mkmmd_loss.py:128]: hmmm...that's true about the single batch. I'd be good with forming XX, YY, XY for batch mkmmd calculations. I think that's a good suggestions.

File: fl4health/losses/mkmmd_loss.py (hunk 2 above)
Comment (emersodb) [reply fl4health/losses/mkmmd_loss.py:239]: Yes, you're right. We should only return this if we're minimizing the type two error. Good call.

Comment (emersodb) [reply fl4health/losses/mkmmd_loss.py:239]: Otherwise, we're still good to make it to the else statement below.

File: fl4health/losses/mkmmd_loss.py (hunk 3 above)
Comment (emersodb) [reply fl4health/losses/mkmmd_loss.py:224]: Yup that's good with me.
